tutorials/provisional/mosaic.py

```diff
@@ -0,0 +1,285 @@
+"""
+=======================================
+Complex and semantic figure composition
+=======================================
+
+.. warning::
+
+   This tutorial documents experimental / provisional API.
+   We are releasing this in v3.3 to get user feedback.  We may
+   make breaking changes in future versions with no warning.
+
+
+Laying out Axes in a Figure in a non uniform grid can be both tedious
+and verbose.  For dense, even grids we have `.Figure.subplots` but for
+more complex layouts, such as Axes that span multiple columns / rows
+of the layout or leave some areas of the Figure blank, you can use
+`.gridspec.GridSpec` (see :doc:`/tutorials/intermediate/gridspec`) or
+manually place your axes.  `.Figure.subplot_mosaic` aims to provide an
+interface to visually lay out your axes (as either ASCII art or nested
+lists) to streamline this process.
+
+This interface naturally supports naming your axes.
+`.Figure.subplot_mosaic` returns a dictionary keyed on the
+labels used to lay out the Figure.  By returning data structures with
+names, it is easier to write plotting code that is independent of the
+Figure layout.
+
+
+This is inspired by a `proposed MEP
+<https://github.com/matplotlib/matplotlib/pull/4384>`__ and the
+`patchwork <https://github.com/thomasp85/patchwork>`__ library for R.
+While we do not implement the operator overloading style, we do
+provide a Pythonic API for specifying (nested) Axes layouts.
+
+"""
+import matplotlib.pyplot as plt
+import numpy as np
+
+
+# Helper function used for visualization in the following examples
+def identify_axes(ax_dict, fontsize=48):
+    """
+    Helper to identify the Axes in the examples below.
+
+    Draws the label in a large font in the center of the Axes.
+
+    Parameters
+    ----------
+    ax_dict : Dict[str, Axes]
+        Mapping between the title / label and the Axes.
+
+    fontsize : int, optional
+        How big the label should be
+    """
+    kw = dict(ha='center', va='center', fontsize=fontsize, color='darkgrey')
+    for k, ax in ax_dict.items():
+        ax.text(0.5, 0.5, k, **kw)
+
```

Comment:
this makes the tutorial really really hard to follow along 'cause it's kinda magiking away the actual plotting w/ the axes. I think this tutorial would benefit from verbosity to make the two methods super clear.

Reply:
I very much disagree on this.  The point of this tutorial is to show how to use the `subplot_mosaic` API to generate layouts of Axes.  By putting the label is huge text in the middle it makes the mapping between the input the `subplot_mosaic` and the layout clear.  Doing actually plotting on the Axes will be distracting from the main goal.

Reply:
I cry and get super-frustrated any time I see functions in the docs 'cause it requires me to keep a call stack in my head. 
Also I think we have cross purposes here. the`identify_axes` is good for showing how `subplot_mosaic` **works** because it prints the layout, yes. It's not great at showing how to **use** `subplot_mosaic` because it abstracts away the usage of the axes which is integral to why I should use this method. What about a combination of identify_axes to put text in the background, but then put plotting on top.

Reply:
I think this should have a simple example at the beginning in the "usual" workflow, *including* actual plot commands.  Then this should be followed up with a simple example using this new method of creating the axes, again with the simple plot commands.  After that, there is no need to keep showing plotting commands for showing the features of the method.

Reply:
Conversely, docs that have _almost_ the same boiler plate repeated multiple times is also very frustrating to use because instead of "and a function is called" you have to mentally do the diff of 15 lines of code while scrolling between the examples to sort out what was actually changed.  By putting the stuff that is the same in a function the signal to noise goes way up.

I'm not really convinced that this change adds anything, but I also don't think it does any harm.  

I re-organized the tutorial to start with subplots + plotting -> list with nice names + plotting -> string short hand -> the extra kwargs ->